pdns/dnsname.cc

```diff
@@ -43,10 +43,10 @@ std::ostream & operator<<(std::ostream &os, const DNSName& d)
 
 DNSName::DNSName(const char* p, size_t length)
 {
-  if(p[0]==0 || (p[0]=='.' && p[1]==0)) {
+  if(length == 0 || p[0]==0 || (p[0]=='.' && p[1]==0)) {
```

Comment:
As a test, I made this constructor private. It does seem to be only called form the other constructors.
- `DNSName(char *p)` uses `strlen()`, so does not handle embedded `\0`s. `length` is s string length.
- `DNSName(const std::string& str) `length` is a string length here as well. Could potentially pass a string with embedded `\0`s. The embedded `\0`s would end up in the labels for the regular case I think. For the backslash case the current test would be incorrect. It is unclear to me what `segmentDNSNameRaw` does with embedded `\0`s,
- The other ct does not call `DNSName(char *, size_t)`.

Reply:
I agree the current API is unclear, especially `DNSName(char *p)`. The other constructors are more explicit, you pass the size of the buffer and special characters are supposed to be escaped, otherwise the contract is violated.
I wonder if we could get rid of the `DNSName(char *p)` constructor..

Reply:
Note that `strlen` was always called before https://github.com/PowerDNS/pdns/commit/beb9fad506e4782fa3cd5db0c484b66493943d8b

Reply:
Changing it to a buffer length (including terminating '\0' instead of a string length (excluding '\0') would change things I believe, as the end condition of the loop assumes a string length.

Reply:
Oh, right, it should be a string length indeed.

Reply:
The handle all calls with `(buf, n)` where `buf` has size `n `the `p[1]` access should be fixed, but the loop should also be double checked if it does not access `buf[n]`, as that would also be an OOB access.

Reply:
My concern was that by sending `"Hey"` without terminating `\0` (like a rust `Vec<u8>` (continuous `u8` array on the heap) exposed as a cpp `char *`) with length like 3, I ended up with out-of-bound read line 49 the `strchr` because `strchr` care for `\0`.

I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.

An eventual possibility would be to make a constructor with a `string_view` and use (or not) the `DNSName(char *, size_t)  made private constructor. It explicit the intends. From my perspective to pass Rust continuous data of u8 it's perfect.

- It avoids making a string copy just for the `\0` if the `memchr` found a consensus
- It spells out the intends to pass a continuous portion of memory ONLY until length to construct a `DNSName`

Reply:
> To handle all calls with (buf, n) where buf has size n the p[1] access should be fixed, but the loop should also be double checked if it does not access buf[n], as that would also be an OOB access.

As far as I can tell the loop is fine, as `iter` is checked against `pend` before being used. I'm not a big fan of using pointer arithmetic for that and would have preferred checking an index, but that should  be OK.

> I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.

I think your use-case is valid, to be clear, and we need to fix/clean up our constructors :)
I 100% agree of taking a `std::string_view` instead of a string, by the way, as it is basically free to construct from a string.

Reply:
> My concern was that by sending `"Hey"` without terminating `\0` (like a rust `Vec<u8>` (continuous `u8` array on the heap) exposed as a cpp `char *`) with length like 3, I ended up with out-of-bound read line 49 the `strchr` because `strchr` care for `\0`.

Yes, that is a real issue. But while reviewing this, a few other issues were spotted.

> I am not a c++ expert, no problems if my suggestion does not fit the problem tell me.
> 
> An eventual possibility would be to make a constructor with a `string_view` and use (or not) the `DNSName(char *, size_t) made private constructor. It explicit the intends. From my perspective to pass Rust continuous data of u8 it's perfect.
> 
>     * It avoids making a string copy just for the `\0` if the `memchr` found a consensus
> 
>     * It spells out the intends to pass a continuous portion of memory ONLY until length to construct a `DNSName`

The handle not-NUL terminated buffers we need to avoid the OOB access with length = 1 and p[0] == '.' something like this: (`p[0]==0 ` is redundant I think)?
```
-  if(length == 0 || p[0]==0 || (p[0]=='.' && p[1]==0)) {
+  if(length == 0 || (length == 1 && p[0]=='.')) {
```

Reply:
> Yes, that is a real issue. But while reviewing this, a few other issues were spotted.

English is not my native tong I agree with the issues spotted. :)
> The handle not-`NULL` terminated buffers we need to avoid the OOB access with `length = 1` and `p[0] == '.'` something like this: (`p[0] == 0` is redundant I think)?
```diff
-  if(length == 0 || p[0]==0 || (p[0]=='.' && p[1]==0)) {
+ if(length == 0 || (length == 1 && p[0]=='.')) {
```

If it's clear and documented that a length is string length I think so.

But i am a little bit rusty with C/C++, I will take some time in the weekend to test-out, and investigate the `string_view` constructor.

I have no "time pressure" nor critical bug on it, it is for avoiding copies and cleaning a little bit the design of the dlso_backend (may have an issue by monday).

Reply:
update; Still not have internet in my new apartment yet, so had some issue to contribute this weekend will find some time today. ;)